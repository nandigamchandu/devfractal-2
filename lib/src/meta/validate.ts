import {
  Array,
  Boolean,
  Date as TDate,
  Number,
  Object as TObject,
  String,
} from 'tcomb'
import * as yup from 'yup'
import {
  ArrayRefinements,
  DateRefinements,
  MT,
  NumberRefinements,
  StringRefinements,
} from './types'

// tslint:disable switch-default

const validateNumberRefinements: (
  r: NumberRefinements,
  value: number,
) => boolean = (r, value) => {
  let s: yup.NumberSchema = yup.number().strict(true)
  switch (r.kind) {
    case 'integer':
      s = s.integer()
      break

    case 'max':
      s = s.max(r.value)
      break

    case 'min':
      s = s.min(r.value)
      break

    case 'positive':
      s = s.positive()
      break

    case 'negative':
      s = s.negative()
  }
  return s.isValidSync(value)
}

const validateStringRefinements: (
  r: StringRefinements,
  value: string,
) => boolean = (r, value) => {
  let s: yup.StringSchema = yup.string().strict(true)
  switch (r.kind) {
    case 'email':
      s = s.email()
      break
    case 'url':
      s = s.url()
      break
    case 'lowercase':
      s = s.lowercase()
      break
    case 'uppercase':
      s = s.uppercase()
      break
    case 'maxLength':
      s = s.max(r.value)
      break
    case 'minLength':
      s = s.min(r.value)
      break
    case 'length':
      s = s.length(r.value)
  }
  return s.isValidSync(value)
}

const validateDateRefinements: (r: DateRefinements, value: Date) => boolean = (
  r,
  value,
) => {
  let s: yup.DateSchema = yup.date()
  switch (r.kind) {
    case 'max':
      s = s.max(r.value)
      break
    case 'min':
      s = s.min(r.value)
  }
  return s.isValidSync(value)
}

const validateArrayRefinements: (
  r: ArrayRefinements,
  value: ReadonlyArray<unknown>,
) => boolean = (r, value) => {
  let s: yup.ArraySchema<unknown> = yup.array()
  switch (r.kind) {
    case 'maxLength':
      s = s.max(r.value)
      break
    case 'minLength':
      s = s.min(r.value)
  }
  return s.isValidSync(value)
}

export const validate: (meta: MT, obj: unknown) => boolean = (meta, obj) => {
  // tslint:disable-next-line:switch-default
  switch (meta.kind) {
    case 'number':
      return (
        Number.is(obj) &&
        (meta.refinements === undefined ||
          meta.refinements.every(r => validateNumberRefinements(r, obj)))
      )

    case 'string':
      return (
        String.is(obj) &&
        (meta.refinements === undefined ||
          meta.refinements.every(r => validateStringRefinements(r, obj)))
      )

    case 'boolean':
      return Boolean.is(obj)

    case 'date':
      return (
        TDate.is(obj) &&
        (meta.refinements === undefined ||
          meta.refinements.every(r => validateDateRefinements(r, obj)))
      )

    case 'enum':
      return String.is(obj) && meta.values.some(e => e === obj)

    case 'array':
      if (!Array.is(obj)) {
        return false
      }
      if (
        !(
          meta.refinements === undefined ||
          meta.refinements.every(e => validateArrayRefinements(e, obj))
        )
      ) {
        return false
      }
      for (const e of obj) {
        if (!validate(meta.of, e)) {
          return false
        }
      }
      return true

    case 'object':
      if (!TObject.is(obj)) {
        return false
      }
      for (const k of Object.keys(obj)) {
        if (!validate(meta.properties[k], obj[k])) {
          return false
        }
      }
      return true
  }
}
