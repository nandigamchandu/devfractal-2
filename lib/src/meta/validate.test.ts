import { ArrayMT, EnumMT, MT, ObjectMT, PrimitiveMT } from './index'
import { isValid } from './validate'

const noEx: PrimitiveMT = { kind: 'number' }
const strEx: PrimitiveMT = { kind: 'string' }
const boolEx: PrimitiveMT = { kind: 'boolean' }
const dateEx: PrimitiveMT = { kind: 'date' }

test('number meta', () => {
  expect(isValid(noEx, 100)).toMatchSnapshot()
  expect(isValid(noEx, '100')).toMatchSnapshot()
})

test('number with refinements', () => {
  const noREx: PrimitiveMT = {
    kind: 'number',
    refinements: [
      { kind: 'integer' },
      { kind: 'min', value: 10 },
      { kind: 'max', value: 20 },
    ],
  }
  expect(isValid(noREx, 15)).toMatchSnapshot()
  expect(isValid(noREx, 5)).toMatchSnapshot()
  expect(isValid(noREx, 25)).toMatchSnapshot()
})

test('str meta', () => {
  expect(isValid(strEx, '100')).toMatchSnapshot()
  expect(isValid(strEx, 100)).toMatchSnapshot()
})

test('str meta with refinements', () => {
  const strREx: PrimitiveMT = {
    kind: 'string',
    refinements: [
      { kind: 'email' },
      { kind: 'lowercase' },
      { kind: 'minStringLength', value: 10 },
      { kind: 'maxStringLength', value: 20 },
    ],
  }
  expect(isValid(strREx, 'foobar@gmail.com')).toMatchSnapshot()
  expect(isValid(strREx, 'foobaR@gmail.com')).toMatchSnapshot()
  expect(isValid(strREx, 'f@g.com')).toMatchSnapshot()
  expect(
    isValid(strREx, 'foooooooooooooooooooooooooooooo@gmai.com'),
  ).toMatchSnapshot()
})

test('bool meta', () => {
  expect(isValid(boolEx, true)).toMatchSnapshot()
  expect(isValid(boolEx, false)).toMatchSnapshot()
  expect(isValid(boolEx, 100)).toMatchSnapshot()
})

test('date meta', () => {
  expect(isValid(dateEx, new Date())).toMatchSnapshot()
  expect(isValid(dateEx, '2000-12-2')).toMatchSnapshot()
})

test('enum meta', () => {
  const enumEx: EnumMT = {
    kind: 'enum',
    name: 'color',
    values: ['red', 'green', 'blue'],
  }
  expect(isValid(enumEx, 100)).toMatchSnapshot()
  expect(isValid(enumEx, 'red')).toMatchSnapshot()
  expect(isValid(enumEx, 'green')).toMatchSnapshot()
  expect(isValid(enumEx, 'blue')).toMatchSnapshot()
  expect(isValid(enumEx, 'GREEN')).toMatchSnapshot()
})

test('array meta', () => {
  const arrNoEx: ArrayMT = { kind: 'array', of: noEx }
  expect(isValid(arrNoEx, [])).toMatchSnapshot()
  expect(isValid(arrNoEx, [10, 20])).toMatchSnapshot()
  expect(isValid(arrNoEx, ['10', '20'])).toMatchSnapshot()
})

test('array meta with refinements', () => {
  const arrNoREx: ArrayMT = {
    kind: 'array',
    of: noEx,
    refinements: [
      { kind: 'maxArrayLength', value: 6 },
      { kind: 'minArrayLength', value: 2 },
    ],
  }
  expect(isValid(arrNoREx, [10, 20])).toMatchSnapshot()
  expect(isValid(arrNoREx, [10, 20, 30, 40, 50, 60, 70])).toMatchSnapshot()
  expect(isValid(arrNoREx, [])).toMatchSnapshot()
})

test('array with differently typed elements', () => {
  const arrStrEx: ArrayMT = { kind: 'array', of: strEx }
  expect(isValid(arrStrEx, [])).toMatchSnapshot()
  expect(isValid(arrStrEx, [10, 20])).toMatchSnapshot()
  expect(isValid(arrStrEx, ['10', '20'])).toMatchSnapshot()

  const arrBoolEx: ArrayMT = { kind: 'array', of: boolEx }
  expect(isValid(arrBoolEx, [])).toMatchSnapshot()
  expect(isValid(arrBoolEx, [true, false])).toMatchSnapshot()
  expect(isValid(arrBoolEx, ['10', '20'])).toMatchSnapshot()

  const arrDateEx: ArrayMT = { kind: 'array', of: dateEx }
  expect(isValid(arrDateEx, [])).toMatchSnapshot()
  expect(isValid(arrDateEx, [new Date(), new Date()])).toMatchSnapshot()
  expect(isValid(arrDateEx, ['10', '20'])).toMatchSnapshot()
})

test('object meta', () => {
  const objEx: ObjectMT = {
    kind: 'object',
    properties: {
      name: strEx,
      price: noEx,
      inStock: boolEx,
    },
  }
  expect(
    isValid(objEx, { name: 'iPhone', price: 700, inStock: true }),
  ).toMatchSnapshot()
  expect(
    isValid(objEx, { name: 'iPhone', price: 700, inStock: 10 }),
  ).toMatchSnapshot()
  expect(
    isValid(objEx, { name: 'iPhone', price: '700', inStock: 10 }),
  ).toMatchSnapshot()
  expect(
    isValid(objEx, { name: 100, price: '700', inStock: 10 }),
  ).toMatchSnapshot()
})

test('complex meta', () => {
  const customerMeta: MT = {
    kind: 'object',
    properties: {
      name: strEx,
      type: {
        kind: 'enum',
        name: 'CustomerType',
        values: ['manager', 'programmer', 'team-leader'],
      },
      addresses: {
        kind: 'array',
        of: {
          kind: 'object',
          properties: {
            city: strEx,
            zip: noEx,
            country: strEx,
          },
        },
      },
    },
  }

  // tslint:disable-next-line:typedef
  const customer = {
    name: 'foobar',
    type: 'team-leader',
    addresses: [
      { city: 'hyderabad', zip: 500018, country: 'india' },
      { city: 'chennai', zip: 500038, country: 'india' },
    ],
  }

  expect(isValid(customerMeta, customer)).toMatchSnapshot()
})
