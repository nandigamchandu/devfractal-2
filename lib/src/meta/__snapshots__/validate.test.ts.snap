// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array meta 1`] = `true`;

exports[`array meta 2`] = `true`;

exports[`array meta 3`] = `false`;

exports[`array meta with refinements 1`] = `true`;

exports[`array meta with refinements 2`] = `false`;

exports[`array meta with refinements 3`] = `false`;

exports[`array with differently typed elements 1`] = `true`;

exports[`array with differently typed elements 2`] = `false`;

exports[`array with differently typed elements 3`] = `true`;

exports[`array with differently typed elements 4`] = `true`;

exports[`array with differently typed elements 5`] = `true`;

exports[`array with differently typed elements 6`] = `false`;

exports[`array with differently typed elements 7`] = `true`;

exports[`array with differently typed elements 8`] = `true`;

exports[`array with differently typed elements 9`] = `false`;

exports[`bool meta 1`] = `true`;

exports[`bool meta 2`] = `true`;

exports[`bool meta 3`] = `false`;

exports[`complex meta 1`] = `true`;

exports[`date meta 1`] = `true`;

exports[`date meta 2`] = `false`;

exports[`enum meta 1`] = `false`;

exports[`enum meta 2`] = `true`;

exports[`enum meta 3`] = `true`;

exports[`enum meta 4`] = `true`;

exports[`enum meta 5`] = `false`;

exports[`number meta 1`] = `true`;

exports[`number meta 2`] = `false`;

exports[`number with refinements 1`] = `true`;

exports[`number with refinements 2`] = `false`;

exports[`number with refinements 3`] = `false`;

exports[`object meta 1`] = `true`;

exports[`object meta 2`] = `false`;

exports[`object meta 3`] = `false`;

exports[`object meta 4`] = `false`;

exports[`str meta 1`] = `true`;

exports[`str meta 2`] = `false`;

exports[`str meta with refinements 1`] = `true`;

exports[`str meta with refinements 2`] = `false`;

exports[`str meta with refinements 3`] = `false`;

exports[`str meta with refinements 4`] = `false`;
